presto-spi/src/main/java/com/facebook/presto/spi/statistics/ComputedStatistics.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi.statistics;
+
+import com.facebook.presto.spi.block.Block;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Objects.requireNonNull;
+
+public class ComputedStatistics
+{
+    private final List<String> groupingColumns;
+    private final List<Block> groupingValues;
+    private final Map<TableStatisticType, Block> tableStatistics;
```

Comment:
There are few reasons behind that:

1. In the future, when we compute advanced statistics, we may need block 
2. I followed the way we pass values when passing the predicates: https://github.com/prestodb/presto/blob/master/presto-spi/src/main/java/com/facebook/presto/spi/predicate/EquatableValueSet.java#L300
3. I believe the less "dynamic" typed something is - the better it is. I think `Block` is still better than `Object`. I provides you at least some context of what the value is.

Reply:
Are there any concerns about memory size for `Block`? We have to store all of these in memory at once in the coordinator. (I'm not suggesting this is a problem, but rather asking if you've considered it or done any back-of-the-napkin calculations)

Reply:
Let's think about the worse case scenario right now

Currently all the statistics are `long` at most. One element LongArrayBlock takes 89 bytes.
At once we can insert 100 partitions at max. 

Not sure what to take as a worse case scenario for the number of columns. Technically tables with even 5000 columns are possible.

In case of 4000 columns the worse case scenario is 4 * 89 * 100 * 5000 = `178000000` ~= `170 MB`
But usually tables don't have that many columns. In case of 100 columns we get `3.4 MB`.

In any cases these objects are short living.

The  amount of data in worse case scenario may grow up, if we ever decided to insert 1000 partitions at once instead of 100. Or if we decided to collect advanced statistics, or MIN/MAX for varchar columns. 

We can do various optimizations. We can implement special type of block, that takes only a single value. We can avoid materializing the `computedStatistics` collection. We can change the structure of the `ComputedStatistics` to avoid materialization of `tableStatistics`.

For now i'm a bit reluctant of doing this pre-mature optimization, since even the worth case scenario, `170 MB` is not that bad, giving the lifetime of this memory is a single `getOutput()` invocation.

Reply:
I agree.  A few comments:
> At once we can insert 100 partitions at max.

This is configurable. Exact number will depend on exact needs. eg people doing ETL with Presto may come to a need to bump this value.

> In case of 4000 columns the worse case scenario is 4 * 89 * 100 * 5000 = 178000000 ~= 170 MB

This is final stats. There are also partial stats produced by every writer (which is: `number of nodes * task.writer-count * hive.max-partitions-per-writers * number of non-partition columns` -- `number of nodes * task.writer-count` times more than final stats).
These are being created on worker nodes and send back to the coordinator. I am not sure how memory mgmt on the coordinator works (since there is a ton of unmanaged memory there), but hopefully there is backpressure from remote exchanges preventing overrunning the coordinator in case when partial stats are available roughly at the same time.